### ISSUE TITLE
kernel updates require two reboots

### ISSUE THREAD

**Issue by janxkoci** [elementary/appcenter#2052]
### What Happened?

For some reason - and unique in the Linux world - the new versions of Appcenter require **two** reboots when there are kernel updates in the queue. This is _highly unusual_.

### Steps to Reproduce

1. Open Appcenter to see available updates.
2. If you see the linux kernel under "System updates" - bingo!
3. Hit "Update all" - note that "System updates" has a "Download" button on the right side, rather than "Update".
4. The Appcenter will prompt you to reboot - do it.
5. After a reboot, the Appenter will again show a notification that updates are available.
6. Open Appcenter and see that "System updates" are still there and the linux kernel is still among them.
7. Hit update (again).
8. You will be (again) prompted to reboot. Again, do it.
9. After the second reboot there are _finally_ no more updates.

### Expected Behavior

Kernel updates have always required only one reboot. It's a nice tradition to keep.

### OS Version

7.x (Horus)

### Software Version

Latest release (I have run all updates)

**Comment by janxkoci** [comment]
## updating with `apt` does _not_ require double-reboot

Today, AppCenter crashed while updating (I've seen both linux kernel and also `appcenter` in the list of updates, the second _could_ be responsible for the crash), so I decided to try updating from CLI (first `flatpak update`, then `sudo apt update && sudo apt upgrade`). Later, I rebooted my laptop and opened AppCenter. It did **_not_** bother me about second reboot!

So this is clearly something AppCenter does on its own for some reason. _One major problem with the current behaviour is that **AppCenter holds all system updates until reboot**!_ (see e.g. https://github.com/elementary/files/issues/1858#issuecomment-1700843098)

### speculation
One reason could be that AppCenter cannot update GRUB? I always see some progress bar during reboots after kernel updates, with a short reboot at the end of the progress, before the machine starts to boot into my system. Is this somehow equivalent to what `apt` does at the end of kernel updates (updating GRUB and such)? Can AppCenter also do it then, rather than during next reboot?

**Comment by danirabbit** [comment]
This isn't an AppCenter specific issue. This is because updates are installed offline and affects pretty much all offline update mechanisms on Linux.

Here's a relevant issue report in PackageKit: https://github.com/PackageKit/PackageKit/issues/510

There's been some progress recently upstream about resolving this, I believe in systemd. I can't find the information at the moment but this is an underlying Linux architecture issue and not something that can be solved in AppCenter

**Comment by janxkoci** [comment]
To clarify: this issue is about the current version of AppCenter included in OS 7 _behaving differently_ then older versions of AppCenter, as well all update mechanisms I know from Ubuntu in the past 10+ years. It all works as I'm used to, except the AppCenter in OS 7, which does things differently.

**Comment by janxkoci** [comment]
To clarify: this issue is about the current version of AppCenter included in OS 7 _behaving differently_ then older versions of AppCenter, as well all update mechanisms I know from Ubuntu in the past 10+ years. It all works as I'm used to, except the AppCenter in OS 7, which does things differently.

PS: in the terminology of the linked issue, elementary OS 7 does _three_ reboots to apply a kernel update. So there is still one more than there needs to be.

**Comment by danirabbit** [comment]
@janxkoci Yes, to clarify: AppCenter now installs all system updates offline so the behavior will be different than installing updates via apt while online. Offline updates require an extra reboot but the trade off is a more stable updates process with a lower chance of failure

**Comment by janxkoci** [comment]
@danirabbit Ok, interesting, but the link still only talks about two reboots, while I experience 3-4 reboots (I need to count them properly next time using their definition).

Also, online updates seem to be rock-solid - the same Ubuntu MATE laptop has only 32gb eMMC storage and with 9gb in Dropbox and a few apps I usually have only some 1.2 - 1.5gb of free disk space (below 1gb Dropbox stops syncing). I often observe that updates - especially kernel updates - regularly run out of space in the middle (as in `df -h` reporting `0B` left on main partition), but `apt` still manages to unpack everything and apply the updates. It's quite impressive tbh. (I usually `autoremove` and `autoclean` right after that.)

**Comment by danirabbit** [comment]
Ah here we go, here's a talk about Soft Reboot: https://www.youtube.com/watch?v=5HO1PAcsM7g It looks like this landed in Systemd in July, so maybe in OS 8 we'll be able to have only a single reboot when installing updates

**Comment by danirabbit** [comment]
Yes, all system updates regardless of whether or not they contain kernel updates require a restart because they are installed offline. The only updates that install in the session are Flatpak updates.

We only push out a set of updates about once a month, unless you're running early access, so you should plan to restart your system once a month to stay up to date.

If you're saying you're rebooting all the way into the desktop multiple times then something there is wrong because what you should see is a reboot, installation of updates happening offline, another reboot, and then your updated desktop

**Comment by janxkoci** [comment]
Well I see system updates definitely more often than once a month, but some of those will be Ubuntu updates.

> If you're saying you're rebooting all the way into the desktop multiple times then something there is wrong because what you should see is a reboot, installation of updates happening offline, another reboot, and then your updated desktop

And that's exactly what this issue is about.
